File: exs/mundo_1/c#/004.cs
// faça um programa que leia algo pelo teclado e mostre na tela o seu tipo primitivo e todas as informações possíveis sobre ele

/*
AVISO:

Para a resolução dessa atividade foram usados métodos complexos, ela serve de referências futuras. Não se assuste caso não entenda.
Conhecimentos recomendados: Laços de repetição, estrutura condicional, arrays e seus métodos.
*/

using System;

static class io {
    // entrada de dados
    public static string input(string text) {
        Console.Write(text);
        return Console.ReadLine();
    }

    // saida de dados
    public static void print(object data, string end="\n") {
        string text = data.ToString();

        Console.Write($"{text}{end}");
    }
}

class CheckString {
  private string text;

  // método construtor
  public CheckString(string text) {
    this.text = text;
  }

  // checa o tipo
  public Type type() {
    return this.text.GetType();
  }

  // checa se há apenas números inteiros
  public bool isNumeric() {
    int digitCounter = 0;
    foreach (char letter in this.text) {
      if (char.IsDigit(letter)) {
        digitCounter += 1;
      }
    }

    return (digitCounter == this.text.Length);
  }

  // checa se há apenas letras do alfabeto
  public bool isAlpha() {
    int letterCounter = 0;
    foreach (char letter in this.text) {
      if (char.IsLetter(letter)) {
        letterCounter += 1;
      }
    }

    return (letterCounter == this.text.Length);
  }

  // checa se há letras ou números
  public bool isAlnum() {
    int letterOrDigitCounter = 0;
    foreach (char letter in this.text) {
      if (char.IsLetterOrDigit(letter)) {
        letterOrDigitCounter += 1;
      }
    }

    return (letterOrDigitCounter == this.text.Length);
  }

  // checa se há apenas espaços
  public bool isSpace() {
    int spaceCounter = 0;
    foreach (char letter in this.text) {
      if (char.IsWhiteSpace(letter)) {
        spaceCounter += 1;
      }
    }

    return (spaceCounter == this.text.Length);
  }
}

class Program {
  public static void Main () {
    string text = io.input("Digite algo: ");

    CheckString checkedText = new CheckString(text);

    io.print(text.Length);
    io.print("O tipo é " + checkedText.type());
    io.print("É um número inteiro? " + checkedText.isNumeric());
    io.print("São apenas letras? " + checkedText.isAlpha());
    io.print("São apenas letras ou números? " + checkedText.isAlnum());
    io.print("São apenas espaços? " + checkedText.isSpace());
    // maiusculo
    // minusculo
    // capitalizado
    // titulo
  }
}

File: exs/mundo_1/javascript/004.js
/*
Desafio 004

Problema: Faça um programa que leia algo pelo teclado e mostre na tela
          seu tipo primitivo e todas as informações possíveis sobre ela.

Resolução do problema:
*/

/*
AVISO:

Para a resolução dessa atividade foram usados métodos complexos, ela serve de referências futuras. Não se assuste caso não entenda.
Conhecimentos recomendados: Laços de repetição, estrutura condicional, arrays e seus métodos.
*/

const input = require('readline-sync').question
const print = console.log


class checkString {
  constructor(text) {
    this.body = {}
    this.body.text = text
    this.body.array = this.body.text.split('')
  }

  type() {
    return typeof(this.body.text)
  }

  isSpace() {
    return this.body.array.every(chr => /\s/.test(chr))
  }

  isNumeric() {
    return (!isNaN(Number(this.body.text)) && !this.isSpace())
  }

  isAlpha() {
    return (this.body.array.every(chr => isNaN(Number(chr)))  && !this.isSpace())
  }

  isAlnum() {
    return (
      this.body.array.some(chr => isNaN(Number(chr))) &&
      this.body.array.some(chr => !isNaN(Number(chr))) &&
      !this.isSpace()
    ) 
  }

  isUpper(_default = this.body.text) {
    return ((_default == _default.toUpperCase()) && this.isAlpha() && !this.isSpace())
  }

  isLower(_default = this.body.text) {
    return ((_default == _default.toLowerCase()) && this.isAlpha() && !this.isSpace())
  }

  isCapitalize(_default = this.body.text) {
    return (
      this.isUpper(_default[0]) && 
      this.isLower(_default.slice(1))
    )
  }

  isTitle() {
    let wordList = this.body.text.split(' ')

    return wordList.every(word => this.isCapitalize(word))
  }
}

function yesOrNo(bool) {
  return bool? 'sim' : 'não'
}

let text = input("Digite algo: ")
checker = new checkString(text)

print('O tipo é ' + checker.type())
print('São apenas espaços? ' + yesOrNo(checker.isSpace()))
print('São apenas números? ' + yesOrNo(checker.isNumeric()))
print('São apenas letras? ' + yesOrNo(checker.isAlpha()))
print('São apenas letras e números? ' + yesOrNo(checker.isAlnum()))
print('São apenas maiúsculas? ' + yesOrNo(checker.isUpper()))
print('São apenas minúsculas? ' + yesOrNo(checker.isLower()))
print('É capitalizado? ' + yesOrNo(checker.isCapitalize()))
print('É um título? ' + yesOrNo(checker.isTitle()))